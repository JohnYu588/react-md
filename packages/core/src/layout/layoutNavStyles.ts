import { cnb } from "cnbuilder";

import { bem } from "../utils/bem.js";

const styles = bem("rmd-layout-nav");

/**
 * @since 6.0.0
 */
export interface LayoutNavClassNameOptions {
  className?: string;

  /**
   * Set this to `true` if the nav should appear below the fixed app bar.
   *
   * @defaultValue `false`
   */
  appBarOffset?: boolean;
}

/**
 * @since 6.0.0
 */
export function layoutNav(options: LayoutNavClassNameOptions = {}): string {
  const { className, appBarOffset } = options;

  return cnb(styles({ "offset-v": appBarOffset }), className);
}
