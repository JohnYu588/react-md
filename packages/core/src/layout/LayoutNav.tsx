"use client";

import { cnb } from "cnbuilder";
import { type HTMLAttributes, type ReactNode, forwardRef } from "react";

import {
  type BaseSheetClassNameOptions,
  DEFAULT_SHEET_CLASSNAMES,
  DEFAULT_SHEET_TIMEOUT,
  sheet,
} from "../sheet/styles.js";
import {
  type CSSTransitionClassNames,
  type TransitionActions,
  type TransitionCallbacks,
  type TransitionTimeout,
} from "../transition/types.js";
import { useCSSTransition } from "../transition/useCSSTransition.js";
import { layoutNav } from "./layoutNavStyles.js";

/**
 * @since 6.0.0
 */
export interface LayoutNavProps
  extends HTMLAttributes<HTMLDivElement>,
    BaseSheetClassNameOptions,
    TransitionCallbacks,
    TransitionActions {
  children: ReactNode;

  /**
   * The component to render as.
   *
   * @defaultValue `"nav"`
   */
  as?: "nav" | "div";

  /**
   * Set this to `true` to display the navigation.
   */
  expanded: boolean;

  /**
   * Set this to `true` to force the navigation to appear below the fixed app
   * bar.
   *
   * @defaultValue `false`
   */
  appBarOffset?: boolean;

  /** @defaultValue {@link DEFAULT_SHEET_TIMEOUT} */
  timeout?: TransitionTimeout;
  /** @defaultValue {@link DEFAULT_SHEET_CLASSNAMES} */
  classNames?: CSSTransitionClassNames;
}

/**
 * **Client Component**
 *
 * This component is used to render a persistent navigation on the left of the
 * page. There is some built-in functionality to support resizing with the
 * `LayoutWindowSplitter` and expandable layouts. Do not use this component for
 * temporary navigation and instead use a `Sheet`.
 *
 * @example Full Height Layout
 * ```tsx
 * "use client";
 * import { LayoutAppBar } from "@react-md/core/layout/LayoutAppBar";
 * import { LayoutNav } from "@react-md/core/layout/LayoutNav";
 * import { Main } from "@react-md/core/layout/Main";
 * import { useHorizontalLayoutTransition } from "@react-md/core/layout/useHorizontalLayoutTransition";
 * import type { ReactElement, PropsWithChildren } from "react";
 *
 * function Layout({ children }: PropsWithChildren): ReactElement {
 *   const { elementProps } = useHorizontalLayoutTransition({
 *     transitionIn: true,
 *   });
 *
 *   return (
 *     <>
 *       <LayoutAppBar {...elementProps}>
 *         <YourAppBarContent />
 *       </LayoutAppBar>
 *       <LayoutNav expanded>
 *         <YourNavigationComponent />
 *       </LayoutNav>
 *       <Main navOffset appBarOffset {...elementProps}>
 *         {children}
 *       </Main>
 *     </>
 *   );
 * }
 * ```
 *
 * @example Toggleable Layout
 * ```tsx
 * "use client";
 * import { LayoutAppBar } from "@react-md/core/layout/LayoutAppBar";
 * import { Main } from "@react-md/core/layout/Main";
 * import { useHorizontalLayoutTransition } from "@react-md/core/layout/useHorizontalLayoutTransition";
 * import { useToggle } from "@react-md/core/useToggle";
 * import MenuIcon from "@react-md/material-icons/MenuIcon";
 * import { cnb } from "cnbuilder";
 * import type { ReactElement, PropsWithChildren } from "react";
 *
 * function Layout({ children }: PropsWithChildren): ReactElement {
 *   const { toggled: expanded, toggle } = useToggle();
 *   const { elementProps } = useHorizontalLayoutTransition({
 *     transitionIn: expanded,
 *   });
 *
 *   return (
 *     <>
 *       <LayoutAppBar {...elementProps}>
 *         <Button
 *           aria-label="Navigation"
 *           onClick={toggle}
 *           buttonType="icon"
 *         >
 *           <MenuIcon />
 *         </Button>
 *         <YourAppBarContent />
 *       </LayoutAppBar>
 *       <LayoutNav expanded>
 *         <YourNavigationComponent />
 *       </LayoutNav>
 *       <Main navOffset={expanded} appBarOffset {...elementProps}>
 *         {children}
 *       </Main>
 *     </>
 *   );
 * }
 * ```
 *
 * @since 6.0.0
 */
export const LayoutNav = forwardRef<HTMLDivElement, LayoutNavProps>(
  function LayoutNav(props, ref) {
    const {
      as: Component = "nav",
      "aria-labelledby": ariaLabelledBy,
      "aria-label": ariaLabel = Component == "nav" && !ariaLabelledBy
        ? "Navigation"
        : undefined,
      expanded,
      children,
      className,
      timeout = DEFAULT_SHEET_TIMEOUT,
      classNames = DEFAULT_SHEET_CLASSNAMES,
      appear,
      enter,
      exit,
      onEnter,
      onEntering,
      onEntered,
      onExit,
      onExited,
      onExiting,
      appBarOffset,
      ...remaining
    } = props;
    const { elementProps } = useCSSTransition({
      nodeRef: ref,
      timeout,
      className: cnb(
        layoutNav({ appBarOffset }),
        sheet({
          className,
          horizontalSize: "none",
          disableOverlay: true,
        })
      ),
      classNames,
      enter,
      exit,
      appear,
      onEnter,
      onEntering,
      onEntered,
      onExit,
      onExited,
      onExiting,
      exitedHidden: true,
      transitionIn: expanded,
    });

    return (
      <Component
        aria-label={ariaLabel}
        aria-labelledby={ariaLabelledBy}
        {...remaining}
        {...elementProps}
      >
        {children}
      </Component>
    );
  }
);
