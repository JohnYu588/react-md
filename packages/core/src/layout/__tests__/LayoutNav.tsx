import { describe, expect, it } from "@jest/globals";
import { createRef } from "react";

import { rmdRender, screen } from "../../test-utils/index.js";
import { LayoutNav, type LayoutNavProps } from "../LayoutNav.js";
import { layoutNav } from "../layoutNavStyles.js";

describe("LayoutNav", () => {
  it("should apply the correct styling, HTMLAttributes, and allow a ref", () => {
    const ref = createRef<HTMLDivElement>();
    const props = {
      ref,
      children: "Content",
      expanded: true,
    } as const;

    const { rerender } = rmdRender(<LayoutNav {...props} />);

    const nav = screen.getByRole("navigation", { name: "Navigation" });
    expect(ref.current).toBeInstanceOf(HTMLElement);
    expect(ref.current).toBe(nav);
    expect(nav).toMatchSnapshot();

    rerender(
      <LayoutNav
        {...props}
        style={{ color: "white" }}
        className="custom-class-name"
      />
    );
    expect(nav).toMatchSnapshot();

    rerender(<LayoutNav {...props} as="div" data-testid="div" />);
    expect(nav).not.toBeInTheDocument();
    const div = screen.getByTestId("div");
    expect(ref.current).toBeInstanceOf(HTMLDivElement);
    expect(ref.current).toBe(div);
    expect(div).toMatchSnapshot();
  });

  it("should apply an aria label while rendered as a nav", () => {
    const props: LayoutNavProps = {
      expanded: true,
      children: "Content",
    };
    const { rerender } = rmdRender(<LayoutNav {...props} />);

    const nav = screen.getByRole("navigation", { name: "Navigation" });
    expect(nav).toBeInTheDocument();

    rerender(<LayoutNav {...props} aria-label="Custom" />);
    expect(nav).toBeInTheDocument();
    expect(nav).toHaveAttribute("aria-label", "Custom");

    rerender(<LayoutNav {...props} aria-labelledby="another-id" />);
    expect(nav).toBeInTheDocument();
    expect(nav).not.toHaveAttribute("aria-label");

    const divProps = {
      ...props,
      as: "div" as const,
      "data-testid": "div",
    };
    rerender(<LayoutNav {...divProps} />);

    expect(nav).not.toBeInTheDocument();
    const div = screen.getByTestId("div");
    expect(div).not.toHaveAttribute("aria-label");

    rerender(<LayoutNav {...divProps} aria-label="My Label" />);
    expect(div).toHaveAttribute("aria-label", "My Label");
  });

  describe("styling utility class", () => {
    it("should be callable without any arguments", () => {
      expect(layoutNav()).toMatchSnapshot();
    });
  });
});
