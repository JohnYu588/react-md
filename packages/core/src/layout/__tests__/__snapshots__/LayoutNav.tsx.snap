// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayoutNav should apply the correct styling, HTMLAttributes, and allow a ref 1`] = `
<nav
  aria-label="Navigation"
  class="rmd-layout-nav rmd-sheet rmd-sheet--horizontal rmd-sheet--left"
>
  Content
</nav>
`;

exports[`LayoutNav should apply the correct styling, HTMLAttributes, and allow a ref 2`] = `
<nav
  aria-label="Navigation"
  class="rmd-layout-nav rmd-sheet rmd-sheet--horizontal rmd-sheet--left custom-class-name"
  style="color: white;"
>
  Content
</nav>
`;

exports[`LayoutNav should apply the correct styling, HTMLAttributes, and allow a ref 3`] = `
<div
  class="rmd-layout-nav rmd-sheet rmd-sheet--horizontal rmd-sheet--left"
  data-testid="div"
>
  Content
</div>
`;

exports[`LayoutNav styling utility class should be callable without any arguments 1`] = `"rmd-layout-nav"`;
